File: src/r/R/problem-4.R
#                         Largest palindrome product
#
# A  palindromic  number  reads  the same both ways. The largest palindrome made
# from the product of two 2-digit numbers is 9009 = 91 × 99.
#
# Find the largest palindrome made from the product of two 3-digit numbers.
#
#                             Answer: 906609
#
#' @useDynLib projectEuler
NULL

#' Largest Palindrome Product
#'
#' @description
#' \code{problem4} calculates the largest palindrome from the product of two 
#' n-digit numbers, being 'n' the giving number of digits
#'
#' @section Complexity:
#' This function has the following complexity \textrm{O}(n)
#'
#' @author Fazendaaa
#'
#' @seealso \url{https://projecteuler.net/problem=4} for more info about it
#'
#' @param digits Number of max digits to run comparison
#'
#' @return The result number
#'
#' @examples
#' problem4(1)
#' problem4(2)
#' problem4(3)
#'
#' @export
problem4 <- function(digits) {
    lower <- 10 ** (digits - 1)
    upper <- (10 ** digits) - 1

    return (Reduce(function (acc, cur) maxPalindrome(acc, cur, upper),
            (lower:upper)))
}

File: src/python/src/project_euler/first/problem_4.py
"""
                        Largest palindrome product

A  palindromic  number reads the same both ways. The largest palindrome made
from the product of two 2-digit numbers is 9009 = 91 × 99.

Find the largest palindrome made from the product of two 3-digit numbers.

                            Answer: 906609
"""


def three_digits_palindrome():
    """Function that returns all the possibles aplindromes made from the product
    of two three-digit numbers"""
    palindrome = []

    for index_i in range(100, 1_000):
        for index_ii in range(100, 1_000):
            if str(index_i * index_ii)[::-1] == str(index_i * index_ii):
                palindrome.append(index_i * index_ii)

    return sorted(palindrome)


print(three_digits_palindrome()[-1])
